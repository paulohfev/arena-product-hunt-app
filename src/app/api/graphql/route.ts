import { NextRequest } from 'next/server'

const PRODUCT_HUNT_API_URL = process.env.NEXT_PUBLIC_API_URL
const API_TOKEN = process.env.API_TOKEN

export async function POST(request: NextRequest) {
  const body = await request.json()

  const response = await fetch(PRODUCT_HUNT_API_URL!, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${API_TOKEN}`,
    },
    body: JSON.stringify(body),
  })

  const data = await response.json()
  return Response.json(data)
}
