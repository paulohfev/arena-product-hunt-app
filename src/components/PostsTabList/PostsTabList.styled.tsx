import styled from 'styled-components'

export const TabsControls = styled.div`
  display: flex;
  gap: 16px;
  justify-content: center;
  margin-bottom: 2rem;
  width: 100%;
`

export const TabButton = styled.div<{ $isActive: boolean }>`
  border-bottom: 2px solid ${({ $isActive }) => ($isActive ? 'orange' : 'transparent')};
  color: ${({ $isActive }) => ($isActive ? 'orange' : 'grey')};
  cursor: pointer;
  display: flex;
  justify-content: center;
  font-weight: 600;
  padding: 16px;
  width: 50%;
  transition: all 0.2s;

  &:hover {
    background: ${({ $isActive }) => ($isActive ? 'transparent' : '#e5e5e5')};
  }
`

export const PostsList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
`

export const LoadingSpinner = styled.div`
  text-align: center;
  padding: 1rem;
  color: #666;
`
