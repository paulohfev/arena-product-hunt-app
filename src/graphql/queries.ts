import { gql } from '@apollo/client'

export const GET_POPULAR_POSTS = gql`
  query {
    posts(order: VOTES, first: 10) {
      nodes {
        id
        name
        tagline
        votesCount
        url
        thumbnail {
          url
        }
      }
    }
  }
`
