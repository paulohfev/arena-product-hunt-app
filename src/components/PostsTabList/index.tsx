import { useEffect, useRef, useState } from 'react'

import { useQuery } from '@apollo/client'
import { toast } from 'react-toastify'

import { GET_POSTS } from '@/graphql/queries'
import { Post } from '@/types/Post'
import { PostQueryData } from '@/types/PostQueryData'

import LoadingSpinner from '../LoadingSpinner'
import PostCard from '../PostCard'
import { PostsList, TabButton, TabsControls } from './PostsTabList.styled'

enum TabType {
  POPULAR = 'popular',
  RECENT = 'recent',
}

const POSTS_PER_PAGE = 10

const PostsTabList = () => {
  const [activeTab, setActiveTab] = useState<TabType>(TabType.POPULAR)
  const observerRef = useRef<IntersectionObserver | null>(null)
  const loadMoreRef = useRef<HTMLDivElement>(null)

  const { data, loading, error, fetchMore } = useQuery<PostQueryData>(GET_POSTS, {
    variables: {
      order: activeTab === TabType.POPULAR ? 'VOTES' : 'NEWEST',
      first: POSTS_PER_PAGE,
    },
  })

  useEffect(() => {
    if (error) {
      toast.error('Failed to load posts. Please try again later.')
    }
  }, [error])

  const posts = data?.posts?.nodes || []
  const hasNextPage = data?.posts?.pageInfo.hasNextPage
  const endCursor = data?.posts?.pageInfo.endCursor

  useEffect(() => {
    const observer = new IntersectionObserver(
      (entries) => {
        if (entries[0].isIntersecting && hasNextPage && !loading) {
          fetchMore({
            variables: {
              after: endCursor,
              first: POSTS_PER_PAGE,
            },
            updateQuery: (prev: PostQueryData, { fetchMoreResult }: { fetchMoreResult?: PostQueryData }) => {
              if (!fetchMoreResult) return prev
              return {
                posts: {
                  ...fetchMoreResult.posts,
                  nodes: [...prev.posts.nodes, ...fetchMoreResult.posts.nodes],
                },
              }
            },
          })
        }
      },
      { threshold: 1.0 }
    )

    if (loadMoreRef.current) {
      observer.observe(loadMoreRef.current)
    }

    observerRef.current = observer

    return () => {
      if (observerRef.current) {
        observerRef.current.disconnect()
      }
    }
  }, [fetchMore, hasNextPage, loading, endCursor])

  return (
    <div>
      <TabsControls>
        <TabButton $isActive={activeTab === TabType.POPULAR} onClick={() => setActiveTab(TabType.POPULAR)}>
          Popular
        </TabButton>
        <TabButton $isActive={activeTab === TabType.RECENT} onClick={() => setActiveTab(TabType.RECENT)}>
          Recent
        </TabButton>
      </TabsControls>

      <PostsList>
        {loading ? <LoadingSpinner /> : posts.map((post: Post) => <PostCard key={post.id} post={post} />)}

        <div ref={loadMoreRef} style={{ height: '20px' }}></div>
      </PostsList>
    </div>
  )
}

export default PostsTabList
