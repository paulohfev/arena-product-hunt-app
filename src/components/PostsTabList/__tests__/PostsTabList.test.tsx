import { MockedProvider } from '@apollo/client/testing'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { toast } from 'react-toastify'

import { GET_POSTS } from '@/graphql/queries'

import PostsTabList from '../index'

// Mock react-toastify
jest.mock('react-toastify', () => ({
  toast: {
    error: jest.fn(),
  },
}))

const mockPosts = {
  nodes: [
    {
      id: '1',
      name: 'Test Post 1',
      tagline: 'Test Tagline 1',
      description: 'Test Description 1',
      votesCount: 100,
      url: 'https://example.com/1',
      slug: 'test-post-1',
      thumbnail: {
        url: 'https://example.com/image1.jpg',
      },
    },
    {
      id: '2',
      name: 'Test Post 2',
      tagline: 'Test Tagline 2',
      description: 'Test Description 2',
      votesCount: 50,
      url: 'https://example.com/2',
      slug: 'test-post-2',
      thumbnail: {
        url: 'https://example.com/image2.jpg',
      },
    },
  ],
  pageInfo: {
    hasNextPage: true,
    endCursor: 'cursor1',
  },
}

const popularPostsMock = {
  request: {
    query: GET_POSTS,
    variables: {
      order: 'VOTES',
      first: 10,
    },
  },
  result: {
    data: {
      posts: mockPosts,
    },
  },
}

const recentPostsMock = {
  request: {
    query: GET_POSTS,
    variables: {
      order: 'NEWEST',
      first: 10,
    },
  },
  result: {
    data: {
      posts: mockPosts,
    },
  },
}

const errorMock = {
  request: {
    query: GET_POSTS,
    variables: {
      order: 'VOTES',
      first: 10,
    },
  },
  result: {
    data: {
      posts: mockPosts,
    },
  },
  error: new Error('An error occurred'),
}

const renderComponent = (mocks = [popularPostsMock]) => {
  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <PostsTabList />
    </MockedProvider>
  )
}

describe('PostsTabList', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders loading state initially', () => {
    renderComponent()
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument()
  })

  it('renders posts after loading', async () => {
    renderComponent()

    await waitFor(() => {
      expect(screen.getByText('Test Post 1')).toBeInTheDocument()
      expect(screen.getByText('Test Post 2')).toBeInTheDocument()
    })
  })

  it('shows error toast when query fails', async () => {
    renderComponent([errorMock])

    await waitFor(() => {
      expect(toast.error).toHaveBeenCalledWith('Failed to load posts. Please try again later.')
    })
  })

  it('switches between Popular and Recent tabs', async () => {
    renderComponent([popularPostsMock, recentPostsMock])

    await waitFor(() => {
      expect(screen.getByText('Test Post 1')).toBeInTheDocument()
    })

    fireEvent.click(screen.getByText('Recent'))
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument()

    await waitFor(() => {
      expect(screen.getByText('Test Post 1')).toBeInTheDocument()
    })
  })

  it('matches snapshot', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })
})
