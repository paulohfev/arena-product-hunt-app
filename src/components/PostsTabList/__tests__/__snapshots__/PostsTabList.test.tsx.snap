// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsTabList matches snapshot 1`] = `
<div>
  <div>
    <div
      class="sc-4a2dfd63-0 dxcazj"
    >
      <div
        class="sc-4a2dfd63-1 klQrGv"
      >
        Popular
      </div>
      <div
        class="sc-4a2dfd63-1 cJtVLR"
      >
        Recent
      </div>
    </div>
    <div
      class="sc-4a2dfd63-2 eLvWxQ"
    >
      <img
        alt="loading"
        class="sc-66b55120-0 bzrstx"
        data-nimg="1"
        data-testid="loading-spinner"
        decoding="async"
        height="48"
        loading="lazy"
        src="/_next/image?url=%2Ficons%2Floading.png&w=96&q=75"
        srcset="/_next/image?url=%2Ficons%2Floading.png&w=48&q=75 1x, /_next/image?url=%2Ficons%2Floading.png&w=96&q=75 2x"
        style="color: transparent;"
        width="48"
      />
      <div
        style="height: 20px;"
      />
    </div>
  </div>
</div>
`;
